### PR TITLE
 pretty: try to align exprs at right of keyword, de-indent AND and OR

### PR DESCRIPTION
**Pull request by knz**
This PR adds a new option `Align` in `PrettyCfg` and makes good use of it to right-align SQL keywords if there is enough space on the line.

When enabled, the alignment is performed using a mix of tabs and spaces so that the output remains properly aligned even if the display tab width changes after the output has been generated (requested by @mjibson).

If the option is disabled, or there is not enough space on the line, the formatting uses "nested sections" as before.

For example, before or with `Align` disabled:

```
--------------------------------------------------
SELECT
    count(*) AS count,
    winner,
    counter * 60 * 5 AS counter
FROM
    (
        SELECT
            winner,
            round(length / 60 / 5) AS counter
        FROM players
        WHERE
            build = $1
            AND hero = $2
            OR region = $3
    )
GROUP BY winner, counter;
```

With `Align` enabled:

```
--------------------------------------------------
  SELECT count(*) AS count,
         winner,
         counter * 60 * 5 AS counter
    FROM (
            SELECT winner,
                   round(
                    length / 60 / 5
                   ) AS counter
              FROM players
             WHERE build = $1
               AND hero = $2
                OR region = $3
         )
GROUP BY winner, counter;
```

Note that tabs are still used whenever possible if the `UseTabs` option is set. The value of `UseTabs` has no bearing on the chosen layout.

### CODE REVIEW

**Review by madelynnblue — COMMENTED** [review]
Code LGTM, but I'm not sold on the formatting. Specifically:

1) The Align stuff assumes that a user's editor is displaying tabs at the same width as the pretty printer. When that doesn't happen, things like

```
SELECT 1,
    2
````

can occur where the `1` and `2` aren't visually aligned. If you follow this thought experiment to the end, it suggests we should just remove tabs altogether and use spaces. I think that *if* we decide we want this kind of alignment, that probably makes the most sense.

2) The right alignment removes a nice property we had, which was that all things at conceptual level `N` were indented the same. (For example, `SELECT a FROM t` would always put `a` and `t` at the same indentation.) This change removes that property, and I kinda liked it.

3) Regarding de-indenting AND and OR, this has the unfortunate side effect of bringing those operators to the same visual level (especially with right alignment) as WHERE, ORDER BY, FROM, etc. One of the purposes of indenting (with tabs or spaces) is to create a visual hierarchy. Deindenting some things weakens that since one can no longer immediately tell at which conceptual level a thing is only based on its initial indentation.

Overall: I like tabs but could be convinced to change. I like tabs because it's easy to deal with during manual editing. But if I'm not doing any more manual editing, spaces would be fine. For the others, I am open to suggestions about how to handle them, since they are quite aesthetic and could vary by person. Voting polls in our slack? Half serious.

**Review by knz — COMMENTED** [review]
Thanks for your early feedback. I think there are two topics of discussion here, which I would like to group as follows:

A) one topic about the particular formatting choices (independent of the discussion about tabs vs spaces)

I made a mistake to show you this PR about alignment of expressions and negative nesting for AND/OR before I asked you to look at  #27264. As a matter of fact, in https://github.com/knz/prettier/ I did it the other way - first align the various SELECT sub-clauses, and only then align the operands and only then AND/OR.

The changes in #27264 resolve point (2) of your previous comment.

It does not really make sense to look at this PR before the other one, and I apologize for wasting your time.

Now, how to move forward with that discussion:

A.1) I will soon (perhaps that will wait until Friday) re-organize the PR to present the changes "in the right order" and weave #27264 into it. 
A.2) I will change the PR to make this alignment behavior configurable. (I believe we need to make this configurable so as to enable a discussion between parties to reveal where preferences stand.) Let's call this option `align` for future reference below.
A.3) we'll discuss the particular alignment of AND and OR, to resolve point (3) of your previous comment.

I have a specific argument, due to the way SQL logical planning works, why AND and OR should be negatively indented relative to their operands (I can explain later). Assuming the argument holds, I could however be convinced (maybe) that they can still be collectively indented relative to the sub-clause keyword. So perhaps we can solve your concern (3) with a middle ground around here:

```
WHERE     foo
      AND bar
      AND baz
```

or

```
WHERE
    foo
AND bar
AND baz
```


What do you think? But I suggest we defer that discussion to after points A.1 and A.2 have gotten somewhere.

Then 

B) discussion relative to tabs vs spaces

I'll want to start first with your remark:

> The Align stuff assumes that a user's editor is displaying tabs at the same width as the pretty printer.

To be honest at first I wasn't really comfortable with this remark. It seemed to me that **using tabs for pretty-printing /anything/ only makes sense if the tool used to view the result was using the same width as the pretty-printer**. This is inherently true of any system that /produces/ tabs for alignment. This is also the reason why most projects strongly favor using spaces for everything, since it's hard to guarantee that different people will use the same tab widths and the appearance of text/code files containing tabs is uncontrollable otherwise.

However, on second read, I tried harder to understand what you meant, to find out why you thought that it wasn't important before. Then I discovered that you had written your code to provide that property: it so appears that one can change the tab width *a posteriori* and still get perfectly aligned output given the current implementation of `doc()` methods. 

This property is, for example, important since most people will use a combination of text editors at 4 spaces per tab and terminals at 8 spaces per tab.

You obviously seem to care about that property a lot.

For the record, I do not care as much (I usually visualize my tabs at 8 spaces per tab even in my editors!), but I can acknowledge your concern here.

Meanwhile, I do fantastically care about the readability of SQL queries, and I think that the alignment discussed above in (A) is a very, very clear improvement overall for complex queries. 

So perhaps we can find a middle ground as follows:

- if the `align` parameter is set, then default `useTabs` to `false` always, so that the display is reliable everywhere by default.
- if `align` is not set, default `useTabs` to whatever we agree on later (I'd be ok with defaulting to `true`, but perhaps other people will have other opinions).

What do you think?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27260#-:-LGyn2a15bNcIe4Ptru8:b-278e4l)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Comment by mjibson** [comment]
My understanding of these various PRs is that #27264 should not be reviewed because it's going away and this PR will be fixed in the upcoming future, at which point I should review it again. Is that correct?

A.2) Regarding adding a new `align` option. This makes me sad because I think we're going in the wrong direction (I want fewer options, not more). However I'm ok with adding it if it encourages discussion about preferences and allows us to make opinionated decisions in the future.

A.3) I'm totally ok with

```
WHERE     foo
      AND bar
      AND baz
```

I tried to do something like that myself (but with `foo` on its own line) unsuccessfully.

B) Your "on second read" paragraph is accurate. I believe pretty strongly that if we do allow tabs, we should maintain the property that whatever the tab-width of the editor, it always produces visually aligned code.

If we are going to add an `align` option, I might merge it into the `useTabs` option as an enum with three values: `align`, `tabs`, `spaces`. Or maybe remove `spaces` altogether?

**Comment by knz** [comment]
Yes your understanding is correct, I suggest to continue this discussion after next Friday when I rework both PRs.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by knz** [comment]
Note that this idea to de-indent AND and OR has a rather big precedent; it is used throughout the SQLite  test suite.

For example here: https://www.sqlite.org/sqllogictest/artifact/f9499fdbafa34e1e

**Comment by knz** [comment]
All right I have added the align option to do the AND/OR differently. You can test them with sqlfmt with the PR in https://github.com/mjibson/sqlfmt/pull/19.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27260)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/util/pretty/pretty_test.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27260#-LGtVuxoYjQFDjK4x_bu-r1-30:-LHSx2kc0NR8Sn0MVWlv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c26a4d8bf5e111702fd31314bd2c25444844ab53/pkg/util/pretty/pretty_test.go#L30)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I think `useTabs = false` would be a clearer example here because the example at 15 would be lined up regardless of editor configuration.
</blockquote></details>

Done.
___
*[pkg/util/pretty/util.go, line 222 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/27260#-LGtVuxoYjQFDjK4x_bv-r5-222:-LHSy97c84PuJKt139Ul:bovzcwn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e5d298f5fb0950fd6c29d85fa811d6fb7ec951/pkg/util/pretty/util.go#L222)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I might use `append` instead of `items[i]` here and init items with `make([]Doc, 0, len(rows))` so that nil entries don't appear in the list. It happens to work because Stack calls ConcatLine which does the correct thing on nil, but it feels safer to be correct here.
</blockquote></details>

Done.

I do want to point out here that you have relied on this behavior of Stack() in other places: the pretty-printing of mutation statements. These were using `Stack(node.With.doc(p), ...)` assuming that things would work if the doc() method returned `nil`. 

I have added a commit to this PR to remove this requirement on `Stack()` (by updating the pretty-printing of mutations), then another to remove the special handling of `nil`.
___
*[pkg/util/pretty/util.go, line 223 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/27260#-LGtVuxoYjQFDjK4x_bv-r5-223:-LHSzMvRAejMNOMTEEvn:birs5p5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a1e5d298f5fb0950fd6c29d85fa811d6fb7ec951/pkg/util/pretty/util.go#L223)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

The commits here are not in correct order. The Align, JoinGroupAligned, and NestT functions don't exist yet. This came up because I wanted to check if Align already Group'd its return, making the call to Group here needless. I'm not sure how annoying it is to reorder them, but it might be helpful if it's doable.
</blockquote></details>

Are you sure? I have double-checked and every commit in turn passes all tests.

As to your question: no, `Align` does not group its return value (it shouldn't - that's not its purpose).

File: pkg/sql/sem/tree/pretty_test.go [hunk 1]
```diff
@@ -47,13 +47,18 @@ func TestPrettyData(t *testing.T) {
...
 	t.Run("ref", func(t *testing.T) {
 		runTestPrettyData(t, "ref", cfg, matches)
 	})
-	cfg.Align = true
+	cfg.Align = tree.PrettyAlignAndDeindent
 	t.Run("align", func(t *testing.T) {
 		runTestPrettyData(t, "align", cfg, matches)
 	})
+	cfg.Align = tree.PrettyAlignOnly
+	t.Run("align", func(t *testing.T) {
```
Comment (mjibson) [inline pkg/sql/sem/tree/pretty_test.go]: align -> `align2` probably?

File: pkg/sql/sem/tree/pretty_test.go [hunk 2]
```diff
@@ -47,13 +47,18 @@ func TestPrettyData(t *testing.T) {
...
+		runTestPrettyData(t, "align2", cfg, matches)
```
Comment (mjibson) [inline pkg/sql/sem/tree/pretty_test.go]: `align` and `align2` in the file names are a bit vague. Could we do something descriptive instead? Maybe `deindent` and `align`? Or `align-deindent` and `align-only`? Also I don't care too much so leave as is if you like.

**Review by madelynnblue — APPROVED** [review]

File: pkg/sql/sem/tree/pretty_test.go (hunk 2 above)
Comment (knz) [reply pkg/sql/sem/tree/pretty_test.go]: Went for more descriptive names. Thanks for the suggestion.

**Comment by knz** [comment]
thank you!

bors r+

**Pull request merged**